LocoNet and LocoIO

silverailscolo opened the pull request:
Spent some time on base LocoNet routines (``instance()`` gone, Javadoc, logging, JUnit testing style)
LocoIO decoder definition experiment expanded to ports 1-4 demo (see Issue #5226)
Legacy LocoIO tool improved UI and entry checking. More i18n of Ln tools, Dutch translation.

bobjacobsen commented:
SpotBugs indicated an error:
```
[INFO] Incompatible bit masks in (e & 0x7f == 0x80) yields a constant result in jmri.jmrix.loconet.locoio.LocoIOPanel.addrSet() [jmri.jmrix.loconet.locoio.LocoIOPanel] At LocoIOPanel.java:[line 300] BIT_AND
```
(That bit operation is always false)

devel-bobm commented:
I see several places where delays have been increased to 800 mSec in order to prevent NPE when "the TC isn't initialized yet" on a slot machine.  Changing the delays seems like "squeezing the balloon" - there will always be a "slower" machine out there.  

I'd think it would be better to "catch" the NPE and then perform some sort of "recovery" to maintain the functionality.

silverailscolo commented on java/src/jmri/jmrix/loconet/LnPacketizer.java:
```diff
@@ -107,15 +107,16 @@ public void sendLocoNetMessage(LocoNetMessage m) {
             }
+        } catch (NullPointerException npe) {
+            log.warn("xmtHandler.notify() npe");
+            throw new NullPointerException(); // caught by LnSensor/PowerManager run()
```
@devel-bobm is this how you suggested to catch the npe?

silverailscolo commented:
Ready for review, intended for 4.11.6.

devel-bobm commented on java/src/jmri/jmrix/loconet/LnPowerManager.java:
```diff
@@ -179,8 +180,22 @@ public void run() {
+                    tc.sendLocoNetMessage(msg);
+                    log.debug("delayed init of LnPowerManager succeeded, sent msg");
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt(); // retain if needed later
```
I am not well-versed in JAVA, so I don't know what effect that "interrupting the current thread" would have.

devel-bobm commented on java/src/jmri/jmrix/loconet/LnPowerManager.java:
```diff
@@ -179,8 +180,22 @@ public void run() {
             msg.setElement(1, 0);
             msg.setElement(2, 0);
-            tc.sendLocoNetMessage(msg);
+            try {
```
This try/catch/try/catch gives a maximum number of two attempts, and will suit any system where 1.5 seconds "is sufficient".

I had assumed a "try, with sufficient delay between tries, until success" for this type of protection against delayed traffic controller initialization...

silverailscolo replied:
Thanks. That calls for [try-catch-finally](http://tutorials.jenkov.com/java-exception-handling/basic-try-catch-finally.html) or ``while``. Will try (!) to construct one.

bobjacobsen commented:
You could put `while (true)` around the initial `try { tc.sendLocoNetMessage(msg);`, and then return on success or if there's a catch of an NPE do a delay the `break;` into another turn of the loop.

Try-finally is really for another purpose.

silverailscolo commented:
Travis headed error:
``[ERROR] Failures: 
[ERROR]   LayoutEditorTest.testSetShowHelpBar:397 getShowHelpBar``
Unrelated, runs OK locally on macOS.

devel-bobm commented:
So far as I know, it is generally not a good idea to have so many different types of changes wrapped up in a single pull-request, especially when the "title" does not reflect the types of changes.  This "pull-request" seems to have become more of a "catch-all" pull-request than a targeted one.

I see functional changes in a variety of places, beyond LocoIO functionality.  I see properties changes in a wide variety of places across a wide variety of LocoNet-based functionality.  I see help file changes.  I see logging changes in code related to sound projects.

I see files which only have formatting changes which do not happen to be "improvements" in the big scheme of things.  I suspect that many seem like they may be "automated" changes.  I wonder how many of these "formatting" changes were put in-place by some other "automated" tool which happened to be configured in a different way.  I don't recall seeing any JMRI-specific recommendations for "formatting tool" settings  recommendations, so cannot comment on what might be "proper". 

_There are currently 306 files which have been touched as part of this pull-request._  That seems excessive to me.  In my _opinion_, this is an out-of-control "pull-request", and either:
. this pull-request needs to be "backed-out" and committed as several different pull-requests, or
. development on this pull-request needs to stop now and the pull-request accepted/rejected as-is, and further development made as new "pull-requests".  

@bobjacobsen, can we get some direction on this?

silverailscolo commented:
I appreciate your asking @devel-bobm. No automated tools or rules were used, just search and my own reading the code. The focus - as the PR title covers - is LocoNet. Just like I did for the tools in the other hardware connection types over the last couple of months, I checked and streamlined the I18N use of properties. As I read the code, I also worked on what caught my eye, mostly routine things like the use of variables in Bundle strings instead of all the ``+`` concatenation (which does not aid the I18N nor speed). Javadoc clean-up helps to get a clean display in the browser, following the documentation.
What did surprise me was the incomplete implementation of the system connection memo/multi-connection support in parts of the LocoNet code. So I fixed the Ctors in line with the other systems (many marked deprecated since a long time, but still in use). My plan was to see the Hexfile LocoNet simulator responses appear in the Ln Monitor, but they still don't. Don't get impressed by the number of files involved (compared to the impressive ``jmrix.loconet`` 30k lines of code). 
To address your last point: My purpose of showing this as a PR 9 days ago was to allow you to see this develop. Since then I only added the Ln tools and changed the wait/retry routine. Unless you spot strange things I hope we can share this in the next dev. release well before May 12. If requested it could be split up in say 3 PR's (no idea yet how IntelliJ lets me do that).

bobjacobsen commented:
@silverailscolo @devel-bobm Different people have different processes.

One thing to avoid is making conflicting and/or inconsistent changes that'll have to be figured out later.  This can happen in space (from one branch to another, when you're working on two projects alternately) or in time (if you start something, set it aside or make it a PR that doesn't get merged for a bit, and then work on something else).  

One way to avoid problems of that kind is to _always_ work on a single branch. That seems what's done here and it's fine.  I generally push smaller PRs and then continue with a new branch

``` 
git checkout -b first-stuff
(changes, create PR)
git checkout -b second-stuff
```
so that while first-stuff is waiting to be merged you can make additional updates and push them to second-stuff (which was branched form your starting point at the end of first-stuff)

My practice is slightly different, but that doesn't mean it's better. I tend to have a "accumulated-changes" branch where I'll make a JavaDoc fix, or add to a help page, or whatever small changes I notice while I'm working.  Those are all independent little things, so I can just push and PR that branch whenever it's got a bunch on it.  Then I have a branch per project where I can keep them separate, each from project.  This works pretty well when I get distracted by my day job and something has to sit around for a while.  But your mileage may differ.

In any case, I'll do a pre-merge look-over on this over shortly.



I tend to have a branch open to make little cleanups, separate from an specific project I'm working on

bobjacobsen commented:
This looks good on a read through. It does make some (overdue) updates to the memo setup, but I don't see anything problematic there. But I'd like to run it on hardware before merging (I get a little nervous about system-specific code like this, as our unit tests aren't all that complete) which I can't do until Sunday. That'll still be before the release.  Thanks for your patience.

bobjacobsen commented:
Thanks for your patience on this
